lib/jumubase_web/views/encoders/ecto_encoders.ex

```diff
@@ -1,18 +1,29 @@
-# The following Poison.Encoder implementations ensure that JSON-encoding an
+# The following Jason.Encoder implementations ensure that JSON-encoding an
 # Ecto changeset or record doesn't result in errors because of un-encodable
 # constructs (such as keyword lists).
 
-defimpl Poison.Encoder, for: Ecto.Changeset do
+require Protocol
+alias Jumubase.Showtime.{Appearance, Participant, Performance, Piece}
+
+# White-list schema fields that may be exposed via JSON
+Protocol.derive(Jason.Encoder, Performance, only: [:id, :contest_category_id, :appearances, :pieces])
+Protocol.derive(Jason.Encoder, Appearance, only: [:id, :instrument, :role, :participant])
+Protocol.derive(Jason.Encoder, Participant, only: [:id, :given_name, :family_name, :birthdate, :phone, :email])
+Protocol.derive(Jason.Encoder, Piece, only: [:id, :title, :composer, :composer_born, :composer_died, :artist, :epoch, :minutes, :seconds])
+
+defimpl Jason.Encoder, for: Ecto.Changeset do
   import JumubaseWeb.ErrorHelpers, only: [translate_error: 1]
 
-  def encode(changeset, options) do
+  def encode(changeset, _options) do
+    data = if changeset.action == :insert, do: %{}, else: changeset.data
```

Comment:
This is a hack to prevent having to encode `Ecto.Association.NotLoaded` values, because insert changesets hold an empty struct (all fields `nil` and no associations loaded) under `data`. Again, not nice.